hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/utils/SparkValueMetadata.java

```diff
@@ -0,0 +1,230 @@
+import org.apache.spark.sql.types.StringType;
+import org.apache.spark.sql.types.TimestampType;
+import org.apache.spark.sql.types.VarcharType;
+
+import java.sql.Date;
+import java.sql.Timestamp;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.util.Arrays;
+
+public class SparkValueMetadata extends ValueMetadata {
+
+  private final DataType dataType;
+
+  protected SparkValueMetadata(ValueType valueType, DataType dataType) {
+    super(valueType);
+    this.dataType = dataType;
+  }
+
+  public Comparable convertSparkToJava(Object value) {
+    return convertSparkToJava(value, true);
+  }
+
+  public static SparkValueMetadata getValueMetadata(DataType dataType, HoodieIndexVersion indexVersion) {
+    if (indexVersion.lowerThan(HoodieIndexVersion.V2)) {
+      return SparkV1EmptyMetadata.get();
+    }
+    if (dataType == null) {
+      return new SparkValueMetadata(ValueType.NULL, null);
+    }
+    ValueType valueType = fromDataType(dataType);
+    if (valueType == ValueType.DECIMAL) {
+      return new SparkDecimalMetadata((DecimalType) dataType);
+    } else {
+      return new SparkValueMetadata(valueType, dataType);
+    }
+  }
+
+  private static ValueType fromDataType(DataType dataType) {
+    return fromDataType(dataType, true);
+  }
+
+  private static ValueType fromDataType(DataType dataType, boolean root) {
+    if (dataType instanceof NullType) {
+      return ValueType.NULL;
+    } else if (dataType instanceof BooleanType) {
+      return ValueType.BOOLEAN;
+    } else if (dataType instanceof IntegerType || dataType instanceof ShortType || dataType instanceof ByteType) {
+      return ValueType.INT;
+    } else if (dataType instanceof LongType) {
+      return ValueType.LONG;
+    } else if (dataType instanceof FloatType) {
+      return ValueType.FLOAT;
+    } else if (dataType instanceof DoubleType) {
+      return ValueType.DOUBLE;
+    } else if (dataType instanceof StringType || dataType instanceof CharType || dataType instanceof VarcharType) {
+      return ValueType.STRING;
+    }  else if (dataType instanceof TimestampType) {
+      return ValueType.TIMESTAMP_MICROS;
+    }  else if (dataType instanceof DecimalType) {
+      return ValueType.DECIMAL;
+    } else if (dataType instanceof DateType) {
+      return ValueType.DATE;
+    } else if (dataType instanceof BinaryType) {
+      return ValueType.BYTES;
+    } else if (dataType instanceof ArrayType) {
+      if (root) {
+        return fromDataType(((ArrayType) dataType).elementType(), false);
+      } else {
+        throw new IllegalArgumentException("Array of Array is not supported");
+      }
+    } else if (SparkAdapterSupport$.MODULE$.sparkAdapter().isTimestampNTZType(dataType)) {
+      return ValueType.LOCAL_TIMESTAMP_MICROS;
+    } else {
+      throw new IllegalArgumentException("Unsupported data type: " + dataType);
+    }
+  }
+
+  private static class SparkV1EmptyMetadata extends SparkValueMetadata {
```

Comment:
Add Javadocs. Is this used for V1 column stats by setting `valueType` as null in the metadata record?

Reply:
removed